### PR TITLE
Add Statistics grouped at Stage ID level in the V2 Engine Response

### PR DESCRIPTION
**Pull request by KKcorps**
Currently, we only return aggregated statistics in the broker response. This already helps out in debugging performance  issues. However,  stats aggregated at stage level will be much more useful because the operations done by leaf stage vs intermediate stage differ a lot.

The PR contains the following changes:

* Add a new **BrokerResponseNativeV2** that extends the existing **BrokerResponseNative** class and will contain Stage level stats
* Add support for **BrokerMetrics** by passing down the table name in **OperatorStats**
* Table names also added for leaf stage stats for better understanding of the output
* For now I am showing only the stats which are filled in the **stageStats** json. All of the stats from the BrokerResponseNative class are supported here.

Some new stats are also planned e.g. **numBlocksIn**, **numBlocksErrored** etc. but those will be taken care in next PR since they touch most of the OperatorClasses.

### Sample

Only **stageStats** is the new key in the response. Rest of the fields are same for  backward compatibility.

**Query** -

```sql
SELECT a.playerName, a.teamID, b.teamName 
FROM baseballStats_OFFLINE AS a
JOIN dimBaseballTeams_OFFLINE AS b
ON a.teamID = b.teamID LIMIT 10
```

**Response** - 

```json
{
   "resultTable":{
      "dataSchema":{
         "columnNames":[
            "playerName",
            "teamID",
            "teamName"
         ],
         "columnDataTypes":[
            "STRING",
            "STRING",
            "STRING"
         ]
      },
      "rows":[
         [
            "Fausto Andres",
            "CIN",
            "Cincinnati Reds/Red Stockings"
         ],
         [
            "Fernando Antonio",
            "HOU",
            "Houston Astros/Colt .45s"
         ],
         [
            "Fernando Antonio",
            "HOU",
            "Houston Astros/Colt .45s"
         ],
         [
            "Fernando Antonio",
            "HOU",
            "Houston Astros/Colt .45s"
         ],
         [
            "Harry Frederick",
            "CLE",
            "Cleveland Indians/Naps/Broncos/Bluebirds/Lake Shores"
         ],
         [
            "Harry Frederick",
            "CLE",
            "Cleveland Indians/Naps/Broncos/Bluebirds/Lake Shores"
         ],
         [
            "William Glenn",
            "DET",
            "Detroit Tigers"
         ],
         [
            "William Glenn",
            "DET",
            "Detroit Tigers"
         ],
         [
            "Paul David",
            "CLE",
            "Cleveland Indians/Naps/Broncos/Bluebirds/Lake Shores"
         ],
         [
            "Albert Julius",
            "CLE",
            "Cleveland Indians/Naps/Broncos/Bluebirds/Lake Shores"
         ]
      ]
   },
   "stageStats":{
      "1":{
         "numBlocks":17,
         "numRows":30,
         "stageExecutionTimeMs":667,
         "operatorIds":[
            "MAILBOX_RECEIVE_3_1_0@127.0.0.1:8421",
            "SORT_3_1_0@127.0.0.1:8444",
            "SORT_3_1_0@127.0.0.1:8442",
            "SORT_3_1_0@127.0.0.1:8443",
            "TRANSFORM_3_1_0@127.0.0.1:8444",
            "TRANSFORM_3_1_0@127.0.0.1:8443",
            "TRANSFORM_3_1_0@127.0.0.1:8442"
         ]
      },
      "2":{
         "numBlocks":39,
         "numRows":47795,
         "stageExecutionTimeMs":631,
         "operatorIds":[
            "MAILBOX_RECEIVE_3_2_0@127.0.0.1:8444",
            "MAILBOX_RECEIVE_3_2_0@127.0.0.1:8443",
            "MAILBOX_RECEIVE_3_2_0@127.0.0.1:8442",
            "SORT_3_2_0@127.0.0.1:8442",
            "HASH_JOIN_3_2_0@127.0.0.1:8444",
            "HASH_JOIN_3_2_0@127.0.0.1:8443",
            "SORT_3_2_0@127.0.0.1:8444",
            "SORT_3_2_0@127.0.0.1:8443",
            "HASH_JOIN_3_2_0@127.0.0.1:8442"
         ]
      },
      "3":{
         "numBlocks":11,
         "numRows":195778,
         "stageExecutionTimeMs":103,
         "numSegmentsQueried":1,
         "numSegmentsProcessed":1,
         "numSegmentsMatched":1,
         "numDocsScanned":97889,
         "numEntriesScannedPostFilter":195778,
         "totalDocs":97889,
         "operatorIds":[
            "MAILBOX_RECEIVE_3_3_0@127.0.0.1:8444",
            "MAILBOX_RECEIVE_3_3_0@127.0.0.1:8443",
            "MAILBOX_RECEIVE_3_3_0@127.0.0.1:8442",
            "LEAF_STAGE_TRANSFER_OPERATOR_3_3_0@127.0.0.1:8443"
         ],
         "tableNames":[
            "baseballStats_OFFLINE"
         ]
      },
      "4":{
         "numBlocks":14,
         "numRows":102,
         "stageExecutionTimeMs":18,
         "numSegmentsQueried":1,
         "numSegmentsProcessed":1,
         "numSegmentsMatched":1,
         "numDocsScanned":51,
         "numEntriesScannedPostFilter":102,
         "totalDocs":51,
         "operatorIds":[
            "MAILBOX_RECEIVE_3_4_0@127.0.0.1:8442",
            "MAILBOX_RECEIVE_3_4_0@127.0.0.1:8444",
            "MAILBOX_RECEIVE_3_4_0@127.0.0.1:8443",
            "LEAF_STAGE_TRANSFER_OPERATOR_3_4_0@127.0.0.1:8442"
         ],
         "tableNames":[
            "dimBaseballTeams_OFFLINE"
         ]
      }
   },
   "exceptions":[
      
   ],
   "numServersQueried":0,
   "numServersResponded":0,
   "numSegmentsQueried":2,
   "numSegmentsProcessed":2,
   "numSegmentsMatched":2,
   "numConsumingSegmentsQueried":0,
   "numConsumingSegmentsProcessed":0,
   "numConsumingSegmentsMatched":0,
   "numDocsScanned":97940,
   "numEntriesScannedInFilter":0,
   "numEntriesScannedPostFilter":195880,
   "numGroupsLimitReached":false,
   "totalDocs":97940,
   "timeUsedMs":109,
   "offlineThreadCpuTimeNs":0,
   "realtimeThreadCpuTimeNs":0,
   "offlineSystemActivitiesCpuTimeNs":0,
   "realtimeSystemActivitiesCpuTimeNs":0,
   "offlineResponseSerializationCpuTimeNs":0,
   "realtimeResponseSerializationCpuTimeNs":0,
   "offlineTotalCpuTimeNs":0,
   "realtimeTotalCpuTimeNs":0,
   "segmentStatistics":[
      
   ],
   "traceInfo":{
      
   },
   "numRowsResultSet":10,
   "minConsumingFreshnessTimeMs":0,
   "numSegmentsPrunedByBroker":0,
   "numSegmentsPrunedByServer":0,
   "numSegmentsPrunedInvalid":0,
   "numSegmentsPrunedByLimit":0,
   "numSegmentsPrunedByValue":0,
   "explainPlanNumEmptyFilterSegments":0,
   "explainPlanNumMatchAllFilterSegments":0
}
```

### CODE REVIEW

**Comment by KKcorps** [comment]
I think **BrokerResponseStats** class can be made an extension of **BrokerResponseNative** class as well. Only differrence will be not having a resultTable plus having additional stage level stats.

**Comment by ankitsultana** [comment]
Excited to see this. @KKcorps : Can you confirm if this also uses `OpChainStats` in any way? I was wondering if we could also track queued time in these stats later.

**Comment by KKcorps** [comment]
> Excited to see this. @KKcorps : Can you confirm if this also uses `OpChainStats` in any way? I was wondering if we could also track queued time in these stats later.

Not right now. But I will include those in the next PR along with some other missing stats.

File: pinot-common/src/main/java/org/apache/pinot/common/response/broker/BrokerResponseNativeV2.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+@JsonPropertyOrder({
+    "resultTable", "stageStats", "exceptions", "numServersQueried", "numServersResponded", "numSegmentsQueried",
+    "numSegmentsProcessed", "numSegmentsMatched", "numConsumingSegmentsQueried", "numConsumingSegmentsProcessed",
+    "numConsumingSegmentsMatched", "numDocsScanned", "numEntriesScannedInFilter", "numEntriesScannedPostFilter",
+    "numGroupsLimitReached", "totalDocs", "timeUsedMs", "offlineThreadCpuTimeNs", "realtimeThreadCpuTimeNs",
+    "offlineSystemActivitiesCpuTimeNs", "realtimeSystemActivitiesCpuTimeNs", "offlineResponseSerializationCpuTimeNs",
+    "realtimeResponseSerializationCpuTimeNs", "offlineTotalCpuTimeNs", "realtimeTotalCpuTimeNs", "segmentStatistics",
+    "traceInfo"
+})
+public class BrokerResponseNativeV2 extends BrokerResponseNative {
```
Comment (walterddr) [inline pinot-common/src/main/java/org/apache/pinot/common/response/broker/BrokerResponseNativeV2.java:48]: can we add a test for V1 and V2 native response type?

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/OperatorStats.java [hunk 2]
```diff
@@ -65,14 +66,18 @@ public void recordRow(int numBlock, int numRows) {
...
+    _executionStats.putAll(executionStats);
   }
 
   public Map<String, String> getExecutionStats() {
-    _executionStats.put(OperatorUtils.NUM_BLOCKS, String.valueOf(_numBlock));
-    _executionStats.put(OperatorUtils.NUM_ROWS, String.valueOf(_numRows));
-    _executionStats.put(OperatorUtils.THREAD_EXECUTION_TIME,
+    _executionStats.putIfAbsent(DataTable.MetadataKey.NUM_BLOCKS.getName(), String.valueOf(_numBlock));
+    _executionStats.putIfAbsent(DataTable.MetadataKey.NUM_ROWS.getName(), String.valueOf(_numRows));
+    _executionStats.putIfAbsent(DataTable.MetadataKey.OPERATOR_EXECUTION_TIME_MS.getName(),
```
Comment (walterddr) [inline pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/OperatorStats.java:80]: what's the reason for putIfAbsent?

Comment (KKcorps) [reply pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/OperatorStats.java:80]: .getExecutionStats is called multiple times but recording stats in the map only once is enough.

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/service/QueryDispatcher.java [hunk 3]
```diff
@@ -147,11 +149,22 @@ public static List<DataBlock> reduceMailboxReceive(MailboxReceiveOperator mailbo
...
+            OperatorStats operatorStats = entry.getValue();
+            ExecutionStatsAggregator rootStatsAggregator = executionStatsAggregatorMap.get(0);
+            ExecutionStatsAggregator stageStatsAggregator = executionStatsAggregatorMap.get(operatorStats.getStageId());
+            if (queryPlan != null) {
+              StageMetadata operatorStageMetadata = queryPlan.getStageMetadataMap().get(operatorStats.getStageId());
+              if (!operatorStageMetadata.getScannedTables().isEmpty()) {
+                operatorStats.recordSingleStat(DataTable.MetadataKey.TABLE.getName(),
+                    Joiner.on("::").join(operatorStageMetadata.getScannedTables()));
+              }
+            }
```
Comment (walterddr) [inline pinot-query-runtime/src/main/java/org/apache/pinot/query/service/QueryDispatcher.java:165]: follow up: refactor this part out as utils

**Review by walterddr — APPROVED** [review]
lgtm overall. please follow up with the TODOs in this PR separately. we still have a lot of clean up to do

File: pinot-common/src/main/java/org/apache/pinot/common/datatable/DataTable.java [hunk 4]
```diff
@@ -102,16 +102,14 @@ enum MetadataValueType {
    */
   enum MetadataKey {
     UNKNOWN(0, "unknown", MetadataValueType.STRING),
-    TABLE(1, "table", MetadataValueType.STRING), // NOTE: this key is only used in PrioritySchedulerTest
+    TABLE(1, "table", MetadataValueType.STRING),
```
[the remark this answers is not in the document]
Comment (KKcorps) [reply pinot-common/src/main/java/org/apache/pinot/common/datatable/DataTable.java:105]: This is dependent on `ExecutionStatsAggregator` not being dependent on field names, plus InstanceResponseBlock also not returning the field names. Both of these need to be changed first, for this to work.

**Pull request merged**